test/k8s/res_limits_test.go

```diff
@@ -0,0 +1,512 @@
+						available := Eventually(func() error {
+							opts := cluster.GetKubectlOptions(TestNamespace)
+							opts.Logger = logger.Discard
+							return k8s.WaitUntilPodAvailableE(cluster.GetTesting(), opts, p.Name, 1, 0)
+						}, "3m", "3s").Should(Succeed())
+
+						wg.Done()
+						if available {
+							printLog("pod %s is now available\n", p.Name)
+						} else {
+							printLog("pod %s failed to become available\n", p.Name)
+							x <- false
+						}
+					}(&pod)
+				}
+				wg.Wait()
+				x <- true
+			}(waitCh)
+
+			go func() {
+				val := <-waitCh
+				if val {
+					printLog("data plane pods are now all available, entering the stabilization sleep")
+					time.Sleep(stabilizationSleep)
+				}
+				ret <- val
+			}()
+		}
+
+		watchControlPlane := func(ctx context.Context, errCh chan<- error) {
+			printLog("monitoring health of control plane pods for at most 10 min\n")
+
+			clientset, err := k8s.GetKubernetesClientFromOptionsE(cluster.GetTesting(), cluster.GetKubectlOptions(Config.KumaNamespace))
+			Expect(err).ToNot(HaveOccurred())
+
+			ctx2, cancel := context.WithTimeout(ctx, 10*time.Minute)
+			watcher, err := clientset.CoreV1().Pods(Config.KumaNamespace).Watch(ctx2, metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app=%s", Config.KumaServiceName),
+			})
+			Expect(err).ToNot(HaveOccurred())
+			secondTicker := time.NewTicker(1 * time.Second)
+
+			go func() {
+				for {
+					select {
+					case <-secondTicker.C:
+						pods, err := k8s.ListPodsE(
+							cluster.GetTesting(),
+							cluster.GetKubectlOptions(Config.KumaNamespace), metav1.ListOptions{
+								LabelSelector: fmt.Sprintf("app=%s", Config.KumaServiceName),
+							})
+						Expect(err).ToNot(HaveOccurred())
+						pod := pods[0]
+						metrics, err := k8s.RunKubectlAndGetOutputE(
+							cluster.GetTesting(),
+							cluster.GetKubectlOptions(Config.KumaNamespace),
+							"exec", pod.Name, "-c", "control-plane", "--", "sh", "-c", "wget -O - http://localhost:5680/metrics")
+						Expect(err).ToNot(HaveOccurred())
+						printLog("control plane metrics: %s\n", metrics)
+					case e := <-watcher.ResultChan():
+						if e.Object == nil {
+							cancel()
+							return
+						}
+
+						pod, ok := e.Object.(*corev1.Pod)
+						if !ok {
+							continue
+						}
+
+						if e.Type == watch.Deleted || e.Type == watch.Added {
+							continue
+						}
+
+						if e.Type == watch.Error {
+							errCh <- errors.New("error watching pod")
+							continue
+						}
+
+						if pod.DeletionTimestamp != nil ||
+							pod.Status.Phase == "Pending" ||
+							k8s.IsPodAvailable(pod) {
+							continue
+						}
+
+						args := []string{"logs", "-c", "control-plane", pod.Name}
+						if hasPodContainerCrashed(pod) {
+							args = append(args, "-p")
+						}
+
+						podLogs, err := k8s.RunKubectlAndGetOutputE(
+							cluster.GetTesting(),
+							cluster.GetKubectlOptions(Config.KumaNamespace),
+							args...)
+						Expect(err).ToNot(HaveOccurred())
```

Comment:
shouldn't `Cluster.GetKumaCPLogs` just work?

Reply:
I didn't use `GetKumaCPLogs` before, thanks for that note. Just had a check, I think it's slightly different here. 
The logic here is monitoring every single pod and dump it's log only if it fails and `GetKumaCPLogs` will dump logs from all CP pods.

Reply:
I'll mark this thread as resolved FTM, feel free to open it again.